examples/rcn.py

```diff
@@ -0,0 +1,392 @@
+    for i, pc in enumerate(pooled_channels):
+        for channel_offset, factor in pooled_channel_weights:
+            ch = (i + channel_offset) % num_orients
+            pos_chan = localized[ch]
+            if factor != 1:
+                pos_chan[pos_chan > 0] *= factor
+            np.maximum(pc, pos_chan, pc)
+    bu_msg = np.array(pooled_channels)
+    return bu_msg
+
+
+# %%
+def initialize_evidences(test_img: np.ndarray) -> Dict:
+    """Computes the initial evidences to the PGMax factor graph given a test image.
+    Args:
+        test_img: The image to run inference on.
+    Returns:
+        evidence_updates: A dictionary containing the initial messages to all the variables in the factor graph.
+    """
+
+    bu_msg = get_bu_msg(test_img)
+
+    evidence_updates = {}
+    for idx in range(frcs.shape[0]):
+        frc = frcs[idx]
+
+        unary_msg = -1 + np.zeros((frc.shape[0], M))
+        for v in range(frc.shape[0]):
+            f, r, c = frc[v, :]
+            evidence = bu_msg[f, r - hps : r + hps + 1, c - hps : c + hps + 1]
+            indices = np.transpose(np.nonzero(evidence > 0))
+
+            for index in indices:
+                r1, c1 = index
+                delta_r, delta_c = r1 - hps, c1 - vps
+
+                index = delta_c + vps + (2 * hps + 1) * (delta_r + hps)
+                unary_msg[v, index] = 1
+
+        evidence_updates[idx] = unary_msg
+    return evidence_updates
+
+
+# %% [markdown]
+# ## 4.2 Run map product inference on all test images
+
+# %%
+run_bp_fn, _, get_beliefs_fn = graph.BP(fg.bp_state, 30)
+scores = np.zeros((len(test_set), frcs.shape[0]))
+map_states_dict = {}
+
+for test_idx in range(len(test_set)):
+    img = test_set[test_idx]
+
+    start = time.time()
+    evidence_updates = initialize_evidences(img)
+    end = time.time()
+    print(f"Initializing evidences took {end-start:.3f} seconds for image {test_idx}.")
+
+    start = end
+    map_states = graph.decode_map_states(
+        get_beliefs_fn(run_bp_fn(evidence_updates=evidence_updates))
+    )
+    end = time.time()
+    print(f"Max product inference took {end-start:.3f} seconds for image {test_idx}.")
+
+    map_states_dict[test_idx] = map_states
+    start = end
+    score = tree_util.tree_multimap(
+        lambda evidence, map: jnp.sum(evidence[jnp.arange(map.shape[0]), map]),
+        evidence_updates,
+        map_states,
+    )
+    for ii in score:
+        scores[test_idx, ii] = score[ii]
+    end = time.time()
+    print(f"Computing scores took {end-start:.3f} seconds for image {test_idx}.")
+
+
+# %% [markdown]
+# # 5. Compute metrics (accuracy)
+
+# %%
+test_preds = train_labels[scores.argmax(axis=1)]
+accuracy = (test_preds == test_labels).sum() / test_labels.shape[0]
+
+print(f"accuracy = {accuracy}")
+
+
+# %% [markdown]
+# # 6. Visualize predictions
+
+# %%
+test_idx = 0
+plt.imshow(test_set[test_idx], cmap="gray")
+
+
+# %% [markdown]
+# ## 6.1 Backtrace of some models on this test image
```

Comment:
I changed the visualization to have backtraces of top model for some test images.
![0](https://user-images.githubusercontent.com/2974401/143890565-fac3cf33-3f76-4250-892f-e282f02bc130.png)
Overlaying seemed a bit non-intuitive. Let me know how this looks.

Reply:
![Screenshot from 2021-11-29 20-01-45](https://user-images.githubusercontent.com/10180197/143983661-19ef548f-28af-4229-be26-ccb0b7831da7.png)
Since we do overlaying all the time not sure how it's non-intuitive. Do you mean the detection doesn't look that good? That is understandable as we are working with a small number of templates.

Reply:
Added overlay.